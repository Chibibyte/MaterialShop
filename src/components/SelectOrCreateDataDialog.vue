<template>
  <!-- v-model="dialog" -->
  <v-dialog :value="value" ref="dialog" @input="e => updateValue(e)" width="unset">
    <h1 class="white pa-3 orange">{{title}}</h1>
    <slot :validate="validate"></slot>
    <v-container>
      <v-btn @click="cancel">
        <span>Cancel</span>
      </v-btn>
      <v-btn @click="confirm" :disabled="!valid">
        <span>Confim</span>
      </v-btn>
    </v-container>
  </v-dialog>
</template>


<script>
/**
 * Business Logic
 *    -dialog for a specific CRUD-Operation on state data
 *
 * Inputs:
 *    -title: dialog title
 *    -value: v-model value for deciding to show/hide the dialog
 *    -cancelFunction: function on cancel: dialog closes automatically
 *    -confirmFunction: function on confirm: dialog closes automatically
 * Outputs:
 */
export default {
  name: "SelectOrCreateDataDialog",
  props: {
    title: {
      type: String,
      default: "TITLE"
    },
    cancelFunction: Function,
    confirmFunction: Function,
    value: {
      type: Boolean,
      default: false
    }
  },
  methods: {
    cancel() {
      if (this.cancelFunction) this.cancelFunction();
      this.closeDialog();
    },
    closeDialog() {
      this.$emit("input", false);
    },
    validate(value) {
      this.valid = value;
    },
    confirm() {
      if (!this.valid) return;
      if (this.confirmFunction) this.confirmFunction();
      this.closeDialog();
    },
    updateValue(e) {
      this.$emit("input", e);
    }
  },
  data() {
    return {
      valid: true
    };
  }
};
</script>