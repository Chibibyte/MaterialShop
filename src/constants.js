export default {
    BUTTONICONS: {
        AUTHENTICATION: 'mdi-login-variant',
        ADDRESS: 'mdi-package-variant',
        BANKACCOUNT: 'mdi-bank',
        SHIPPING: 'mdi-truck',
        PAYMENT: 'mdi-cash',
        SUMMARY: 'mdi-text-box-check-outline',
        CHECKOUT_COMPLETE: 'mdi-check',
        ADD: 'mdi-plus',
        REMOVE: 'mdi-minus',
        CANCEL: 'mdi-window-close',
        CONFIRM: 'mdi-check',
        SHOPPING: 'mdi-shopping',
        SUCCESS: 'mdi-check',
        FAILURE: 'mdi-window-close',
        WARNING: 'mdi-alert-circle-outline'
    }
}