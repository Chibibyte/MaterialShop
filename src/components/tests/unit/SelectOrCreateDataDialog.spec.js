import Vuex from 'vuex'
import { shallowMount, createLocalVue } from '@vue/test-utils';
import SelectOrCreateDataDialog from '@/components/SelectOrCreateDataDialog.vue';
import constants from '@/constants.js'

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.prototype.$constants = constants;

let stubs = [
    'SelectOrCreateDataDialog',
    'v-container',
    'v-dialog',
    'v-btn'
]

let wrapper;
beforeEach(() => {
    wrapper = shallowMount(SelectOrCreateDataDialog, { stubs, localVue });
})

describe('Component', () => {
    test('is instantiated', () => {
        expect(wrapper.isVueInstance).toBeTruthy();
    })

    test('default snapshot', () => {
        expect(wrapper.element).toMatchSnapshot();
    })
})

describe('methods: closeDialog', () => {
    test('default', async () => {
        wrapper.vm.closeDialog();
        await wrapper.vm.$nextTick();
        expect(wrapper.emitted().input).toBeTruthy();
        expect(wrapper.emitted().input[0][0]).toBe(false);
        expect(wrapper.element).toMatchSnapshot();
    })
})

describe('methods: cancel', () => {
    test('default: no cancelFunction', async () => {
        wrapper.vm.cancel();
        await wrapper.vm.$nextTick();
        expect(wrapper.emitted().input).toBeTruthy();
        expect(wrapper.emitted().input[0][0]).toBe(false);
    })

    test('with cancelFunction', async () => {
        await wrapper.setProps({ cancelFunction: jest.fn() });
        wrapper.vm.cancel();
        await wrapper.vm.$nextTick();
        // event emitted
        expect(wrapper.emitted().input).toBeTruthy();
        expect(wrapper.emitted().input[0][0]).toBe(false);
        // cancelFunction called
        expect(wrapper.vm.cancelFunction).toBeTruthy();
        expect(wrapper.vm.cancelFunction.mock.calls).toHaveLength(1);
    })
})

describe('methods: confirm', () => {
    test('default: no confirmFunction', async () => {
        wrapper.vm.confirm();
        await wrapper.vm.$nextTick();
        expect(wrapper.emitted().input).toBeTruthy();
        expect(wrapper.emitted().input[0][0]).toBe(false);
    })

    test('with confirmFunction', async () => {
        await wrapper.setProps({ confirmFunction: jest.fn() });
        wrapper.vm.confirm();
        await wrapper.vm.$nextTick();
        // event emitted
        expect(wrapper.emitted().input).toBeTruthy();
        expect(wrapper.emitted().input[0][0]).toBe(false);
        // confirmFunction called
        expect(wrapper.vm.confirmFunction).toBeTruthy();
        expect(wrapper.vm.confirmFunction.mock.calls).toHaveLength(1);
    })
})


