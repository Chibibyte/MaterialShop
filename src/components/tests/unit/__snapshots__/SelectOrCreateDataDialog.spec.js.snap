// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component default snapshot 1`] = `
<v-dialog-stub
  width="unset"
>
  <h1
    class="white pa-3 orange"
  >
    TITLE
  </h1>
    
  <v-container-stub>
    <v-btn-stub>
      <span>
        Cancel
      </span>
    </v-btn-stub>
     
    <v-btn-stub>
      <span>
        Confim
      </span>
    </v-btn-stub>
  </v-container-stub>
</v-dialog-stub>
`;

exports[`methods: closeDialog default 1`] = `
<v-dialog-stub
  width="unset"
>
  <h1
    class="white pa-3 orange"
  >
    TITLE
  </h1>
    
  <v-container-stub>
    <v-btn-stub>
      <span>
        Cancel
      </span>
    </v-btn-stub>
     
    <v-btn-stub>
      <span>
        Confim
      </span>
    </v-btn-stub>
  </v-container-stub>
</v-dialog-stub>
`;
